cmake/sca/Kconfig

```
@@ -0,0 +1,10 @@
+# Static Code Analysis (SCA) options
+
+# Copyright (c) 2024 Baumer (www.baumer.com)
+# SPDX-License-Identifier: Apache-2.0
+
+menu "Static Code Analysis (SCA) tools"
+
+source "cmake/sca/eclair/Kconfig.eclair"
```

This should safe guarded so that it depends on eclair being enabled.

---

Sorry to ask, but how can this be done in Kconfig wen the only information about the SCA in the build command is -DZEPHYR_SCA_VARIANT=<tool>?

---

we should export the SCA tool in use to the Kconfig invocation so that Kconfig can osource the correct SCA tool config.

That will allow any sca tool to provide additional Kconfigs.

---

do you have a example how to do that or a source to point me to so I can take a look into it and how this can be done? I'm not that deep in the Kconfig mechanisms

---

For sourcing sca KConfigs, we can use similar principle as toolchain, where the toolchain kconfig dir is exported to Kconfig like this:
https://github.com/zephyrproject-rtos/zephyr/blob/2e956c239b9279a7af108c45beb8522dc29b45c3/cmake/modules/kconfig.cmake#L166

and then toolchain specific Kconfig is sourced like this:
https://github.com/zephyrproject-rtos/zephyr/blob/2e956c239b9279a7af108c45beb8522dc29b45c3/Kconfig.zephyr#L52

main difference is that Zephyr toolchain variant is always set (so the osource always expand to something), whereas the SCA variant may be empty, which can risk locating empty setting, so that must be properly handled.

But as the above probably requires a bit more work, then short term, what you can do is to export `ZEPHYR_SCA_VARIANT` to Kconfig, similar to toolchain:
https://github.com/zephyrproject-rtos/zephyr/blob/2e956c239b9279a7af108c45beb8522dc29b45c3/cmake/modules/kconfig.cmake#L165
and then apply this around the sourcing:
```suggestion
if "$(ZEPHYR_SCA_VARIANT)" = "eclair"
source "cmake/sca/eclair/Kconfig.eclair"
endif
```

---

Maybe make this more generic?

```
if "$(ZEPHYR_SCA_VARIANT)" != ""
osource "cmake/sca/$(ZEPHYR_SCA_VARIANT)/Kconfig"
endif
```

---

> Maybe make this more generic?
> 
> ```
> if "$(ZEPHYR_SCA_VARIANT)" != ""
> osource "cmake/sca/$(ZEPHYR_SCA_VARIANT)/Kconfig"
> endif
> ```

That would be the case when resolving the follow up issue [#77018](https://github.com/zephyrproject-rtos/zephyr/issues/77018) on which I'm working already

---

> Maybe make this more generic?
> 
> In `cmake/modules/kconfig.cmake`
> 
> ```cmake
> if(DEFINED ZEPHYR_SCA_VARIANT)
>   set(OPERATION WRITE)
>   foreach(root ${SCA_ROOT})
>     file(${OPERATION} ${KCONFIG_BINARY_DIR}/sca/Kconfig
>          "osource \"${root}/cmake/sca/${ZEPHYR_SCA_VARIANT}/Kconfig\"\n"
>     )
>     set(OPERATION APPEND)
>   endforeach()
> endif()
> ```
> 
> In `cmake/sca/Kconfig`
> 
> ```
> if "$(ZEPHYR_SCA_VARIANT)" != ""
> osource "cmake/sca/$(ZEPHYR_SCA_VARIANT)/Kconfig"
> osource "$(KCONFIG_BINARY_DIR)/sca/Kconfig"
> endif
> ```

The Problem with that updated solution would be that the SCA_ROOT isn't defined at that point therfore the foreach wouldn't be executed. Also I try to get rid of the `cmake/sca/Kconfig` file and do it more like the toolchain where you would define the sca and its Kconfig

---

I tried it locally, which seemed to work fine?

---

Ok, did you define sca_root upfront in the environment? Because in my test right now I rely on the default SCA_ROOT which is set to ZEPHYR_BASE

---

I added an `sca_root` to a module:

```yml
name: <some module>
build:
  settings:
    sca_root: .

```

---

Ok but than the case wouldn't work if someone rely on the default SCA_ROOT set internally to ZEPHYR_BASE like I did.
Will try to solve that case aswell, or do you think setting the sca_root should be a requirement and without it it shouldn't work?
But in that case we would need to generate a message to inform the user that he needs to define it

---

You shouldn't set the default `SCA_ROOT`, I think. Simply `osource` them independently.

```
if "$(ZEPHYR_SCA_VARIANT)" != ""
osource "cmake/sca/$(ZEPHYR_SCA_VARIANT)/Kconfig"
osource "$(KCONFIG_BINARY_DIR)/sca/Kconfig"
endif
```
The generated `$(KCONFIG_BINARY_DIR)/sca/Kconfig` will contain all modules that have set `sca_root`.